ampd-handlers/src/bin/stellar/handler.rs

```diff
@@ -0,0 +1,578 @@
+    ) -> Vote {
+        match self {
+            PollEventData::Message(message) => {
+                verify_message(source_gateway_address, tx_receipt, message)
+            }
+            PollEventData::VerifierSet(verifier_set) => {
+                verify_verifier_set(source_gateway_address, tx_receipt, verifier_set)
+            }
+        }
+    }
+}
+
+impl From<PollStartedEvent> for voting::PollStartedEvent<PollEventData, ScAddress> {
+    fn from(event: PollStartedEvent) -> Self {
+        match event {
+            PollStartedEvent::Messages(message_event) => voting::PollStartedEvent {
+                poll_data: message_event
+                    .messages
+                    .into_iter()
+                    .map(PollEventData::Message)
+                    .collect(),
+                poll_id: message_event.poll_id,
+                source_chain: message_event.source_chain,
+                source_gateway_address: message_event.source_gateway_address,
+                expires_at: message_event.expires_at,
+                confirmation_height: message_event.confirmation_height,
+                participants: message_event.participants,
+            },
+            PollStartedEvent::VerifierSet(verifier_set_event) => voting::PollStartedEvent {
+                poll_data: vec![PollEventData::VerifierSet(verifier_set_event.verifier_set)],
+                poll_id: verifier_set_event.poll_id,
+                source_chain: verifier_set_event.source_chain,
+                source_gateway_address: verifier_set_event.source_gateway_address,
+                expires_at: verifier_set_event.expires_at,
+                confirmation_height: verifier_set_event.confirmation_height,
+                participants: verifier_set_event.participants,
+            },
+        }
+    }
+}
+
+#[derive(Debug, TypedBuilder)]
+pub struct Handler<C>
+where
+    C: StellarClient,
+{
+    pub verifier: AccountId,
+    pub voting_verifier_contract: AccountId,
+    pub chain: ChainName,
+    pub rpc_client: C,
+    pub monitoring_client: monitoring::Client,
+}
+
+#[async_trait]
+impl<C> VotingHandler for Handler<C>
+where
+    C: StellarClient + Send + Sync,
+{
+    type Digest = Hash;
+    type Receipt = TxResponse;
+    type ChainAddress = ScAddress;
+    type EventData = PollEventData;
+
+    fn chain(&self) -> &ChainName {
+        &self.chain
+    }
+
+    fn verifier(&self) -> &AccountId {
+        &self.verifier
+    }
+
+    fn voting_verifier_contract(&self) -> &AccountId {
+        &self.voting_verifier_contract
+    }
+
+    fn monitoring_client(&self) -> &monitoring::Client {
+        &self.monitoring_client
+    }
+
+    async fn finalized_txs(
+        &self,
+        poll_data: &[Self::EventData],
+        _confirmation_height: Option<u64>,
+    ) -> Result<HashMap<Self::Digest, Self::Receipt>> {
+        let tx_hashes = poll_data
+            .iter()
+            .map(|data| data.message_id().tx_hash_as_hex_no_prefix().to_string())
```

Comment:
if you change the `tx_hash` function, I think you could do:

```suggestion
            .map(|data| data.tx_hash().to_string())
```